Endpoint for creating a node

samredai opened the pull request:
### Summary

This adds a POST `/nodes/` endpoint that allows for creating a node through the API. The functionality here aims to be the same as what runs through the compile step that processes the yaml files (with the plan being to eventually replace the compile command with calls to the REST API.

I've removed the flags above and switched to the idea of having an environment attribute on the node to signal if it's in a staging environment or a production environment. The node data model now has an "environment" attribute with an enum value of "staging" or "production". I've summarized behaviors for the different validations below:

### Inferring Column Types
- `"environment": "published"` - The request will fail if it can't properly infer the columns.
- `"environment": "draft"` - The server will try to infer the columns and if it can't, it will still allow adding the node.

### Unknown Dimension Tag
I've left this out entirely so right now you can't add dimension tags when creating a node. I started to feel unsure about how this would play with the automatic inferring of column types. If it's ok with everyone, I think it'd be better to tackle this in a follow-up PR to focus in on this. Right now I'm leaning towards having a separate API call that adds a dimension tag to a column in an existing node and we can include this validation logic in that route.

### Auto Discover Source Node Columns
I've removed the ability to add tables when creating a source node (thanks @betodealmeida for highlighting to me that those are optional!). If everyone agrees I think it'd be great to keep it as a separate API call to attach a source node to a table since that feels more involved than the generic "node creation" that the `POST '/nodes/'` route should handle.

### Test Plan

- Ran `docker-compose up`
- Went to the swagger UI at [http:localhost:8000/docs](http:localhost:8000/docs)
- Submitted the following payloads to the `POST /nodes/` route

_request_
```json
{
   "name":"revenue",
   "description":"This is a source table containing revenue data about the product",
   "type":"source",
   "tables":{
      "postgres":[
         {
            "catalog":null,
            "schema_":"product",
            "table":"revenue"
         }
      ]
   },
   "columns":{
      "country":{
         "type":"STR"
      },
      "revenue": {
         "type": "FLOAT"
      }
   }
}
```
_response_
```json
{
  "name": "revenue",
  "description": "This is a source table containing revenue data about the product",
  "type": "source",
  "query": null,
  "id": 13
}
```
_request_
```json
{
   "name":"purchases_over_a_grand",
   "description":"This filters the revenue source node to only include purchases over $1,000",
   "type":"transform",
   "query": "SELECT country, revenue FROM revenue WHERE revenue > 1000.00"
}
```
_response_
```json
{
  "name": "purchases_over_a_grand",
  "description": "This filters the revenue source node to only include purchases over $1,000",
  "type": "transform",
  "query": "SELECT country, revenue FROM revenue WHERE revenue > 1000.00",
  "id": 14,
  "columns": [
    {
      "id": 68,
      "dimension_id": null,
      "dimension_column": null,
      "name": "country",
      "type": "STR"
    },
    {
      "id": 69,
      "dimension_id": null,
      "dimension_column": null,
      "name": "revenue",
      "type": "FLOAT"
    }
  ],
  "tables": []
}
```
_request_
```json
{
   "name":"total_revenue_from_purchases_over_a_grand",
   "description":"A total of product revenue where only purchases over $1,000 are included",
   "type":"metric",
   "query": "SELECT SUM(revenue) as total_revenue FROM purchases_over_a_grand"
}
```
_response_
```json
{
  "name": "total_revenue_from_purchases_over_a_grand",
  "description": "A total of product revenue where only purchases over $1,000 are included",
  "type": "metric",
  "query": "SELECT SUM(revenue) as total_revenue FROM purchases_over_a_grand",
  "id": 15,
  "columns": [
    {
      "id": 70,
      "dimension_id": null,
      "dimension_column": null,
      "name": "total_revenue",
      "type": "FLOAT"
    }
  ],
  "tables": []
}
```
- [x] PR has an associated issue: #255
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

<!-- Any special instructions around deployment? -->

samredai commented:
cc: @betodealmeida @agorajek @CircArgs

agorajek reviewed (commented):
Makes sense to me. One naming comment.

samredai commented on dj/api/metrics.py:
```diff
@@ -143,3 +144,21 @@ async def read_metrics_sql(
         sql=create_query.submitted_query,
     )
+
+@router.post("/metrics/", response_model=MetricRead)
```
I started with something similar, I think it was `/create/metric/` but I thought it looked very clean and intuitive to see the `GET /metrics/` and `POST /metrics/` right next to each other in the swagger UI. It's a recognized pattern to just rely on the HTTP verbs but I'm mostly impartial here. I'll let others chime in as well.

samredai replied:
After the discussion I'm going to try to use a union from the openapi spec and try a single `POST /nodes/` for all node types.

samredai commented:
@CircArgs @betodealmeida @agorajek I updated this to use a single `POST /nodes/` route after our discussion and it seems to work nicely. I still have the different create models (`MetricCreate`, `DimensionCreate`) but fastapi recognizes the `Union` of them.

Currently it only supports creating a metric or dimension but once we settle on a design I'll include source and transform nodes. Here's how the schema shows up in the swagger UI.

<img width="991" alt="Screen Shot 2022-12-03 at 6 31 05 PM" src="https://user-images.githubusercontent.com/43911210/205466536-7eef4296-402a-4c09-91cc-4e74af8fcb5d.png">

agorajek commented on dj/api/nodes.py:
```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+    """
+    dj_settings = get_settings()
+
+    config = {
```
Should we add a marker for the client provided source nodes, say `mock: true`. This setting would prevent the DJ Watchdog from reporting on these source nodes as errors, we could still check them but not wake anyone up if a mock source node is not matching the underlying databases/tables?

samredai replied:
That's a good point. I'll need to think about this some more but I added it as an item to the weekly sync. Allowing CRUD through the API changes a few things that we'd have to consider, for example we will no longer have the signal of file changes to know when DJ should perform a validation analysis on the DAG.

We also have to think about how we handle situations where an upstream node is deleted or modified, making a once valid node now invalid (so if it has a `mock:false` attribute we'd need to know to switch it to `mock:true`). We could run the validation across the entire DAG each time a CRUD API request completes but that seems like the extreme solution and won't scale nicely. This is a tricky item with some nuances but a big one so it's worth all of us brainstorming a ton on it.

betodealmeida replied:
I think ideally we want to run the validation across the DAG on every modification. Why do you think this wouldn't scale, @samredai? The validation can just track column lineage, so it should be fast and not require any database calls.

samredai replied:
That might just be an incorrect assumption on my part but I imagined it's more than just upstream column lineage, right? If we allow for draft/incomplete/mock nodes, then we'd have to check all of them every time a node is added to see if those nodes are now valid. We would even have to do this when a source node is added which has no upstream nodes. So I'm imagining what that would look like for a large DAG with many of these incomplete nodes.

samredai replied:
I've updated the PR with the idea from our discussion earlier to have a staging environment which has softer validation requirements for adding nodes. That punts the discussion in this thread to one of the next PRs when we add the logic to move a node from the staging environment to the production environment and have to perform the full validation. I like how much better this turned out, thanks @betodealmeida @agorajek @CircArgs!

samredai commented on dj/models/source.py:
```diff
@@ -0,0 +1,45 @@
+from dj.typing import ColumnType
+
+
+class SourceCreateTable(TypedDict, total=False):
```
Yeah you're right, I'm scratching my head trying to figure out what the naming represents, haha. I got it from the SQLModel docs that explain how to have [multiple models](https://sqlmodel.tiangolo.com/tutorial/fastapi/multiple-models/#multiple-models-with-inheritance) to represent the full path through a fastapi CRUD request. I'll update them to your suggestions!

samredai replied:
Ok how does this look? I tweaked some of your suggestions a tiny bit.
- SourceCreateTable -> SourceTableSchema
- SourceCreateColumn -> SourceColumnSchema
- SourceCreate -> SourceNodeCreator
- SourceRead -> ExistingSourceNode
- NodeCreate -> NodeCreator
- NodeRead -> ExistingNode

agorajek replied:
*Creator and Existing* makes sense. But I would not be afraid of longer class names for clarity so for the first two items i propose:
- SourceCreateTable -> SourceNodeTableSchema
- SourceCreateColumn -> SourceNodeColumnType

samredai replied:
Updated! Much clearer now, thanks @agorajek

samredai commented on dj/api/nodes.py:
```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+
+
+@router.post("/nodes/", response_model=Union[SourceRead, NodeRead])
+def create_a_node(
```
Updated! 👍

betodealmeida commented:
> * **FAIL_ON_UNKNOWN_DIMENSION_TAG** - If true, raises an error and prevents adding a node if one of the nodes columns are tagged with a dimension for which no dimension node currently exists.
> 
> * **FAIL_ON_INFER_COLUMNS_ERROR** - If true, raises an error and prevents adding a node if any of the columns cannot be inferred from upstream nodes, i.e. the upstream node doesn't exist.

Would we even want this to be false? That could lead to an inconsistent DAG.

betodealmeida commented on dj/api/nodes.py:
```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+    # Check the columns for any dimension references and do a metadata lookup of the dimension node
+    for column in new_node.columns:
+        if dimension_target := column.get("dimension"):
+            dimension_target, dimension_column = dimension_target.rsplit(".", 1)
```
I wonder if we should use a different separator when indicating dimension, eg:

```yaml
dimension = foo.bar::column
```

samredai replied:
Ooh I really like that! I wonder if that would change anything about how the the parsing or converting the parse tree to a DJ AST works.

agorajek replied:
This syntax (::) reminds me of the Postgres notation for casting a value to a type. What exactly do we want to represent here?

samredai replied:
Ah that's right, I knew that looked familiar. 😄

betodealmeida replied:
Doesn't have to be `::`, could be `:`?

betodealmeida replied:
`foo.bar->column`
`foo.bar:column`

betodealmeida commented on dj/api/nodes.py:
```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+                )
+            else:
+                # Use explicitly defined columns if auto-discover is disabled
+                columns = explicitly_defined_columns
```
Another thing we might want to consider here is allowing "patching" the columns, ie, "read the schema information from the database but for this particular column use this type". Maybe have an `overrides` parameter or something like that? Something to discuss.

samredai replied:
+1, I like it!

agorajek replied:
I think we need to clarify the cases of creating source nodes that have nothing to do with reality. I am always for making the system more approachable but I can't think of a good use case yet. Is it basically letting people draw in it like in a scratchpad?

betodealmeida commented on dj/api/nodes.py:
```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+                # Use explicitly defined columns if auto-discover is disabled
+                columns = explicitly_defined_columns
+            config["columns"] = columns
+            config["tables"].append(  # type: ignore
```
I'm confused here, you're adding this table to all nodes, not just those that don't have a database. But also... shouldn't we enforce that all tables have a known database? What's the use case here?

samredai replied:
Ah thanks for reminding me to add this as a discussion topic. Right now this is just a placeholder.

There are two use cases I think are worth supporting here, both related to this idea of allowing people to step into a full DJ deployment. The first is allowing someone to create nodes and generate an ANSI SQL query when calling the `/metrics/` route, using nothing but the DJ server (no external databases). This would be very useful for tutorials, demos, and a way to get familiar with the core concepts around DJ nodes and how they interact with each other. The second is if someone wants to only leverage the ANSI SQL building capabilities of DJ, they can rely on just sending requests to the `/metrics/{node_id}/sql/` route without having to connect all of their source nodes to a database. They could always do that instead by importing the library into a python runtime and using it directly, but it would be nice to facilitate that as an optional mode of the REST API.

samredai commented:
@betodealmeida I definitely see the case for fixing the DAG validations and reflection to always be on, but I feel there's value in providing a path for "legging in", especially in the spirit of making the project more approachable.

- crawling...
  - FAIL_ON_UNKNOWN_DIMENSION_TAG = **False**
  - FAIL_ON_INFER_COLUMNS_ERROR = **False**
  - AUTO_DISCOVER_SOURCE_NODE_COLUMNS = **False**
- walking...
  - FAIL_ON_UNKNOWN_DIMENSION_TAG = **True**
  - FAIL_ON_INFER_COLUMNS_ERROR = **False**
  - AUTO_DISCOVER_SOURCE_NODE_COLUMNS = **False**
- jogging...
  - FAIL_ON_UNKNOWN_DIMENSION_TAG = **True**
  - FAIL_ON_INFER_COLUMNS_ERROR = **True**
  - AUTO_DISCOVER_SOURCE_NODE_COLUMNS = **False**
- _running!_
  - FAIL_ON_UNKNOWN_DIMENSION_TAG = **True**
  - FAIL_ON_INFER_COLUMNS_ERROR = **True**
  - AUTO_DISCOVER_SOURCE_NODE_COLUMNS = **True**

What do you think?

samredai commented on dj/api/nodes.py:
```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+                columns = explicitly_defined_columns
+            config["columns"] = columns
+            config["tables"].append(  # type: ignore
+                Table(
+                    database=Database(
+                        name="unknown",
+                        description=(
+                            "A placeholder database for source "
+                            "nodes with no known database"
+                        ),
+                        URI="unknown://localhost:9999/unknown",
+                    ),
+                    columns=columns,
+                    **table_data,
+                ),
+            )
```
Woah, thanks @betodealmeida! That totally worked. I don't know what made me think tables are required for source nodes. This along with the idea of having a staging environment feels very clean. I'm going to update this PR to do that and remove the global configs to tweak validation.

samredai commented:
@CircArgs @betodealmeida @agorajek please have another look when you have a chance. I've updated the PR as well as the description to summarize the expected behavior.

agorajek commented:
> I've removed the ability to add tables when creating a source node

@samredai  about that... i think you are talking here about the internals rather than user perspective, but just wanted to confirm, because I feel like "registering a table in DJ" would really be equivalent to creating a source node for it and I am not sure if I want to have these two steps separated from the user side. I understand that there may be a need for an empty/staging source node but most of the time I think "adding a table to DJ" should be part of "creating a source node". No?

agorajek reviewed (approved):
LGTM. Just one question on the source node creation logic.

samredai commented:
> @samredai about that... i think you are talking here about the internals rather than user perspective, but just wanted to confirm, because I feel like "registering a table in DJ" would really be equivalent to creating a source node for it and I am not sure if I want to have these two steps separated from the user side. I understand that there may be a need for an empty/staging source node but most of the time I think "adding a table to DJ" should be part of "creating a source node". No?

@agorajek that's a fair point but I don't think making them separate API routes necessarily means it has to always be a two step action for the user. Whatever command the user calls (CLI command, UI click, etc) can make a sequence of calls to create the source node and immediately attach one or more tables to it. I'm not strongly tied to this idea but something doesn't feel right about everything in the `POST /nodes/` being very lightweight DJ metadata operations, but then when the user includes a table it triggers an if statement that makes calls to external databases that can fail for a variety of reasons.

agorajek commented:
@samredai yes, I think I am now convinced that we can start making source nodes w/o any table pointer being necessary right away.

samredai commented:
Rebased and squashed. Keeping this PR open since it needs some more changes given some recent discussions. We may after all want to track the "valid"/"invalid" state of a node with respect to the DAG in addition to keeping "production"/"staging" as functionally meaningful attributes. Looking forward to discussing this more in the new year. Happy holidays!
